crates/vhost-device-gpio/src/backend.rs

```diff
@@ -181,47 +213,21 @@ fn start_backend<D: 'static + GpioDevice + Send + Sync>(args: GpioArgs) -> Resul
             // threads, and so the code uses unwrap() instead. The panic on a thread won't cause
             // trouble to other threads/guests or the main() function and should be safe for the
             // daemon.
-            let controller = match cfg {
-                GpioDeviceType::PhysicalDevice(device_num) => match D::open(device_num) {
-                    Ok(device) => match GpioController::<D>::new(device) {
-                        Ok(controller) => controller,
-                        Err(error) => return Err(Error::CouldNotCreateGpioController(error)),
-                    },
-                    Err(error) => return Err(Error::CouldNotOpenDevice(error)),
-                },
-                _ => return Err(Error::DeviceUnimplemented),
-            };
-
-            let backend = Arc::new(RwLock::new(
-                VhostUserGpioBackend::new(controller).map_err(Error::CouldNotCreateBackend)?,
-            ));
-            let listener = Listener::new(socket.clone(), true).unwrap();
-
-            let mut daemon = VhostUserDaemon::new(
-                String::from("vhost-device-gpio-backend"),
-                backend.clone(),
-                GuestMemoryAtomic::new(GuestMemoryMmap::new()),
-            )
-            .map_err(Error::CouldNotCreateDaemon)?;
-
-            daemon.start(listener).unwrap();
-
-            match daemon.wait() {
-                Ok(()) => {
-                    info!("Stopping cleanly.");
+            match cfg {
+                GpioDeviceType::PhysicalDevice(device_num) => {
+                    let controller = PhysDevice::open(device_num).unwrap();
+                    start_device_backend(controller, socket.clone());
                 }
-                Err(vhost_user_backend::Error::HandleRequest(
-                    vhost_user::Error::PartialMessage | vhost_user::Error::Disconnected,
-                )) => {
-                    info!("vhost-user connection closed with partial message. If the VM is shutting down, this is expected behavior; otherwise, it might be a bug.");
+                #[cfg(feature = "mock_gpio")]
+                GpioDeviceType::SimulatedDevice(num_gpios) => {
+                    let controller = MockGpioDevice::open(num_gpios).unwrap();
+                    start_device_backend(controller, socket.clone());
                 }
-                Err(e) => {
-                    warn!("Error running daemon: {:?}", e);
+                #[cfg(not(feature = "mock_gpio"))]
+                _ => {
+                    panic!("mock implementatoins aren't supported");
                 }
-            }
-
-            // No matter the result, we need to shut down the worker thread.
-            backend.read().unwrap().exit_event.write(1).unwrap();
+            };
```

Comment:
While this change simplifies things a great deal, I would still move the actual construction of the backend with the different feature checks to a separate function.

Reply:
Is this comment stale now?